krun/test_prog.c

```diff
@@ -0,0 +1,159 @@
+    u_int64_t *mperfs, u_int64_t *ctr1s);
+u_int64_t **alloc_array(int n_cores);
+void print_arrays(int n_cores, u_int64_t **aperfs, u_int64_t **mperfs, u_int64_t **ctr1s);
+int get_ctr1_width(void);
+void mask_ctr1s(int n_cores, u_int64_t **ctr1s);
+
+/* ctr1 is variable size */
+u_int64_t ctr1_mask = 0;
+
+int get_ctr1_width(void) {
+	uint32_t eax, edx;
+	int width;
+
+	asm volatile(
+	    "mov $0xa, %%eax\n\t"       // pctr leaf
+	    "cpuid\n\t"
+	    : "=a" (eax), "=d" (edx)    // out
+	    :                           // in
+	    : "ebx", "ecx");            // clobber
+	width = (edx & 0x1fe0) >> 5;
+	printf("ctr1 width is %d\n", width);
+	return width;
+}
+
+void read_msrs_checked(int n_cores, bool ctr1_first, u_int64_t *aperfs,
+    u_int64_t *mperfs, u_int64_t *ctr1s)
+{
+	int err = syscall(__NR_krun_read_msrs, n_cores, ctr1_first, aperfs, mperfs, ctr1s);
+	if (err != 0) {
+		printf("syscall returned: %ld\n", err);
+		exit(err);
+    }
+}
+
+/* allocates space for 2 readings of one performance counter for all cores */
+u_int64_t **alloc_array(int n_cores)
+{
+	u_int64_t **arr;
+	int i;
+
+	arr = calloc(2, sizeof(u_int64_t *));
+	if (arr == NULL) {
+		printf("failed to allocate\n");
+		exit(EXIT_FAILURE);
+	}
+
+	for (i = 0; i < 2; i++) {
+		arr[i] = calloc(n_cores, sizeof(u_int64_t));
+		if (arr[i] == NULL) {
+			printf("failed to allocate\n");
+			exit(EXIT_FAILURE);
+		}
+	}
+	return arr;
+}
+
+/* apply masking to ctr1s */
+void mask_ctr1s(int n_cores, u_int64_t **ctr1s)
+{
+	int core;
+
+	for (core = 0; core < n_cores; core++) {
+		ctr1s[0][core] &= ctr1_mask;
+		ctr1s[1][core] &= ctr1_mask;
+	}
+}
+
+/* print before and after readings for all counters */
+void print_arrays(int n_cores, u_int64_t **aperfs, u_int64_t **mperfs, u_int64_t **ctr1s)
+{
+	int core, idx;
+
+	for (idx = 0; idx < 2; idx++) {
+		if (!idx) {
+			printf("Before:\n");
+		} else {
+			printf("After:\n");
+		}
+		for (core = 0; core < n_cores; core++) {
+			printf("  core: %02d: aperf: %016" PRIu64 "    ", core, aperfs[idx][core]);
+			printf("mperf: %016" PRIu64 "    ", mperfs[idx][core]);
+			printf("ctr1 : %016" PRIu64 "\n", ctr1s[idx][core]);
+
+			// check they make sense too
+			if (aperfs[0][core] > aperfs[1][core]) {
+				errx(1, "bad aperfs");
+			}
+			if (mperfs[0][core] > mperfs[1][core]) {
+				errx(1, "bad mperfs");
+			}
+			if (ctr1s[0][core] > ctr1s[1][core]) {
+				errx(1, "bad ctr1");
+			}
+		}
+	}
+}
+
+int main(void)
+{
+	u_int64_t **aperfs, **mperfs, **ctr1s;
+	int core, i, n_cores;
+
+	ctr1_mask = ((u_int64_t) 1 << get_ctr1_width()) - 1;
+	printf("ctr1 mask is %" PRIx64 "\n", ctr1_mask);
+	n_cores = sysconf(_SC_NPROCESSORS_ONLN);
+
+	/* make space for readings */
+	aperfs = alloc_array(n_cores);
+	mperfs = alloc_array(n_cores);
+	ctr1s = alloc_array(n_cores);
+
+	printf("configuring for %d cores\n", n_cores);
+	syscall(__NR_krun_configure, n_cores);
+	syscall(__NR_krun_reset_msrs, n_cores);
+
+	// continually read MSRs
+	while (true) {
```

Comment:
I'm not sure that an infinitely-running test program is ideal. Or should we maybe print out something like "Checking cores -- press Ctrl-C to quit" so that users know that's expected behaviour?